sklearn_numba_dpex/common/random.py

```diff
@@ -0,0 +1,309 @@
+
+
+def get_random_raw(states):
+    """Returns one `uint64` random integer"""
+    result = dpt.empty(sh=(1,), dtype=np.uint64, device=states.device)
+    make_random_raw_kernel()(states, result)
+    return result
+
+
+@lru_cache
+def make_random_raw_kernel():
+    @dpex.kernel
+    def _get_random_raw_kernel(states, result):
+        result[zero_idx] = _xoroshiro128pp_next(states, zero_idx)
+
+    return _get_random_raw_kernel[1, 1]
+
+
+@lru_cache
+def make_rand_uniform_kernel_func(dtype):
+    """Instanciate a kernel function that returns a random float in [0, 1)"""
+    if not hasattr(dtype, "name"):
+        raise ValueError(
+            "dtype is expected to have an attribute 'name', like np.dtype or numba "
+            "types."
+        )
+
+    if dtype.name == "float64":
+        convert_rshift = uint32(11)
+        convert_const = float64(uint64(1) << uint32(53))
+        convert_const_one = float64(1)
+
+        @dpex.func
+        def uint64_to_unit_float(x):
+            """Convert uint64 to float64 value in the range [0.0, 1.0)"""
+            return float64(x >> convert_rshift) * (convert_const_one / convert_const)
+
+    elif dtype.name == "float32":
+        convert_rshift = uint32(40)
+        convert_const = float32(uint32(1) << uint32(24))
+        convert_const_one = float32(1)
+
+        @dpex.func
+        def uint64_to_unit_float(x):
+            """Convert uint64 to float32 value in the range [0.0, 1.0)
+            NB: this is different than original numba.cuda.random code. Instead of
+            generating a float64 random number before casting it to float32, a float32
+            number is generated from uint64 without intermediate float64. This enables
+            compatibility with devices that do not support float64 numbers.
+            However is seems to be exactly equivalent e.g it passes the float precision
+            test in sklearn.
+            """
+            return float32(x >> convert_rshift) * (convert_const_one / convert_const)
+
+    else:
+        raise ValueError(
+            "Expected dtype.name in {float32, float64} but got "
+            f"dtype.name == {dtype.name}"
+        )
+
+    @dpex.func
+    def xoroshiro128pp_uniform(states, index):
+        """Return one random float in [0, 1)"""
+        return uint64_to_unit_float(_xoroshiro128pp_next(states, index))
+
+    return xoroshiro128pp_uniform
+
+
+def create_xoroshiro128pp_states(n_states, seed=None, subsequence_start=0, device=None):
+    """Returns a new device array initialized for n random number generators.
+
+    This initializes the RNG states so that states in the array correspond to
+    subsequences separated by 2**64 steps from each other in the main sequence.
+    Therefore, as long as no thread requests more than 2**64 random numbers, all the
+    RNG states produced by this function are guaranteed to be independent.
+
+    Parameters
+    ----------
+    n_states : int
+        number of RNG states to create
```

Comment:
Could you please explain how to chose this value? The tests only use `n_states=1` if I am not mistaken.

I guess it has to do with the amount of parallelism you want to allow when generating a large number of random numbers?

Reply:
> I guess it has to do with the amount of parallelism you want to allow when generating a large number of random numbers?

yes. n_states = n_threads

Reply:
Please make that explicit in the docstring for this parameter.

Reply:
The number of threads [is 2**64 ](https://prng.di.unimi.it/)so we should be fine :grin: 

![image](https://user-images.githubusercontent.com/29153872/201730797-052bc9bf-e6ee-417c-aeb9-752c08decf9b.png)

Reply:
About the scope of this PR the purpose here is not to provide an high-level interface dedicated to user that just generate random number like e.g `numpy.random.rand` would do. For me it was only about serving the need of the  kmeans++ where the device functions (those defined with `@dpex.func`) are used in to have rng in each thread fused with other instructions. It is a lower level use where it is up to the user to choose carefully `n_states` and then use it accordingly in other kernels. And we're not preventing bad use, like having several threads using the same state. Maybe I should have made it clearer in the description of the PR so you can start with a better understanding a priori.

(also, there is rng with better user interface, e.g in jax each thread use a different hash function on integers)

That being said I think your docstring is fine and we can even omit to document the constraint of not having more than `2**64` threads iterate on more than `2**64` values since those are virtually impossible to reach ?

> You mean the maximum number of threads that we can use safely?
yes, the algorithm is safe as long as `n_states < 2**64` and each state not updated more than `2**64` times.